jetty-client/src/main/java/org/eclipse/jetty/client/GZIPContentDecoder.java

```diff
@@ -39,13 +47,47 @@ public GZIPContentDecoder(ByteBufferPool byteBufferPool, int bufferSize)
         super(byteBufferPool, bufferSize);
     }
 
+    @Override
+    public void beforeDecoding(HttpExchange exchange)
+    {
+        HttpFields.Mutable headers = exchange.getResponse().headers();
+        // Content-Length is not valid anymore while we are decoding.
+        headers.remove(HttpHeader.CONTENT_LENGTH);
+        // Content-Encoding should be removed/modified as the content will be decoded.
+        ListIterator<HttpField> fields = headers.listIterator();
+        while (fields.hasNext())
+        {
+            HttpField field = fields.next();
+            if (field.getHeader() == HttpHeader.CONTENT_ENCODING)
+            {
+                String value = Arrays.stream(field.getValues())
+                    .filter(encoding -> !"gzip".equalsIgnoreCase(encoding))
+                    .collect(Collectors.joining(", "));
+                if (value.isEmpty())
+                    fields.remove();
+                else
+                    fields.set(new HttpField(HttpHeader.CONTENT_ENCODING, value));
+                break;
+            }
+        }
+    }
+
     @Override
     protected boolean decodedChunk(ByteBuffer chunk)
     {
+        decodedLength += chunk.remaining();
         super.decodedChunk(chunk);
         return true;
     }
 
+    @Override
+    public void afterDecoding(HttpExchange exchange)
+    {
+        HttpFields.Mutable headers = exchange.getResponse().headers();
+        headers.remove(HttpHeader.TRANSFER_ENCODING);
+        headers.putLongField(HttpHeader.CONTENT_LENGTH, decodedLength);
```

Comment:
Is the length always known?  Don't we have the option of streaming decoded content?

Reply:
We have the option of streaming content, and the content length won't be seen until the whole content has been streamed.

This is to support non-streaming use cases, where the response content is buffered, the whole response received, and at that point if the application wants to look at the content length as a header, rather than asking to the buffered content, it'll get the right value.